Add TiFlash disaggregate architecture and S3 support (#12949)
pingcap/docs: tiflash/tiflash-disaggregated-and-s3.md

    1. Open the TiDB configuration file in edit mode:

          ```shell
          tiup cluster edit-config mycluster
          ```

    2. Add the following configuration items to the TiDB configuration file:

        ```shell
        server_configs:
        tidb:
        disaggregated-tiflash: true   # Query TiFlash using the disaggregated storage and compute architecture
        ```

    3. Restart TiDB:

        ```shell
        tiup cluster reload mycluster -R tidb
        ```

## Restrictions

- TiFlash does not support in-place switching between the **disaggregated storage and compute architecture** and the **coupled storage and compute architecture**. Before switching architectures, you must remove all existing TiFlash nodes.
- After the migration from one architecture to another, all TiFlash data needs to be replicated again.
- Only TiFlash nodes with the same architecture are allowed in the same TiDB cluster. Two architectures cannot coexist in one cluster.
- The disaggregated storage and compute architecture only supports object storage using the S3 API, while the coupled storage and compute architecture only supports local storage.

> - When using S3 storage, TiFlash nodes cannot obtain the keys of files not on their own nodes, so the [static encryption](/encryption-at-rest.md) feature is cannot be used.

Author: ```suggestion
- When using S3 storage, TiFlash nodes cannot obtain the keys of files not on their own nodes, so the [Encryption at Rest](/encryption-at-rest.md) feature is cannot be used.
```

- The demand for compute resources has obvious peaks and valleys. For example, intensive reconciliation queries are usually performed at night, which demands high compute resources. In this case, you can consider temporarily adding more Compute Nodes at night. While at other times, you only need fewer Compute Nodes to complete regular query tasks.

## Prerequisites

1. Prepare an Amazon S3 bucket for storing TiFlash data.

    You can also use an existing bucket, but you need to create a dedicated storage directory for each TiDB cluster. For more information about S3 buckets, see [AWS documentation](https://docs.aws.amazon.com/en_us/AmazonS3/latest/userguide/creating-buckets-s3.html).

    You can also use other S3-compatible object storage, such as [MinIO](https://min.io/).

    The S3 APIs used by TiFlash are as follows:

    - PutObject
    - GetObject
    - CopyObject
    - DeleteObject
    - ListObjectV2
    - GetObjectTagging
    - PutBucketLifecycle

2. Add a [lifecycle](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lifecycle-mgmt.html) to the prepared S3 bucket for cleaning up deleted data:

    ```shell
    "Expiration": {
        "Days": 1
    },

>     "Tags": [
>         {
>             "Value": "tiflash_deleted", 
>             "Key": "true"
>         }
>     ]

Reviewer: ```suggestion
"Filter": {
    "And": {
        "Tags": [
            {
                "Value": "tiflash_deleted", 
                "Key": "true"
            }
        ]
    }
}
```

- The demand for compute resources has obvious peaks and valleys. For example, intensive reconciliation queries are usually performed at night, which demands high compute resources. In this case, you can consider temporarily adding more Compute Nodes at night. While at other times, you only need fewer Compute Nodes to complete regular query tasks.

## Prerequisites

1. Prepare an Amazon S3 bucket for storing TiFlash data.

    You can also use an existing bucket, but you need to create a dedicated storage directory for each TiDB cluster. For more information about S3 buckets, see [AWS documentation](https://docs.aws.amazon.com/en_us/AmazonS3/latest/userguide/creating-buckets-s3.html).

    You can also use other S3-compatible object storage, such as [MinIO](https://min.io/).

    The S3 APIs used by TiFlash are as follows:

    - PutObject
    - GetObject
    - CopyObject
    - DeleteObject
    - ListObjectV2
    - GetObjectTagging
    - PutBucketLifecycle

2. Add a [lifecycle](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lifecycle-mgmt.html) to the prepared S3 bucket for cleaning up deleted data:

    ```shell

>     "Expiration": {

Reviewer: ```suggestion
    "ID": "tiflash-clean",
    "Expiration": {
```

## Architecture overview

![TiFlash Write and Compute Separation Architecture](/media/tiflash/tiflash-s3.png)

In the disaggregated storage and compute architecture, different functionalities of the TiFlash process are divided and allocated to two types of nodes: the Write Node and the Compute Node. These two types of nodes can be deployed separately and scaled independently, which means that you can decide the number of Write Nodes and Compute Nodes to be deployed as needed.

- TiFlash Write Node

    The Write Node receives Raft logs data from TiKV, converts the data into the columnar format, and periodically packages and uploads all the updated data within a certain period to S3. In addition, the Write Node manages the data on S3, such as continuously organizing data to improve query performance and deleting useless data.

    The Write Node uses local disks (usually NVMe SSDs) to cache the latest written data to avoid excessive use of memory.

>     The Write Node has a faster scaling speed than the TiFlash node in the coupled storage and compute architecture. That is, after adding or removing Write Nodes, data can reach balance faster among Write Nodes. The reason is that the Write Node stores all the data in S3 and only needs to store a small amount of data locally at runtime. The scaling up and down process is essentially the migration of Region Peers among nodes. To migrate a Region Peer from one Write Node to another Write Node for management, the target Write node only needs to download a small amount of metadata about this Region from the latest files uploaded to S3 by the Region Peer's original Write Node, and then synchronize the latest Region updates from TiKV to catch up with the Region Leader's progress.

Reviewer 2: ```suggestion
    The Write Node has a faster scaling speed than the TiFlash node in the coupled storage and compute architecture. That is, after adding or removing Write Nodes, data can reach balance faster among Write Nodes. The reason is that the Write Node stores all the data in S3 and only needs to store a small amount of data locally at runtime. The scaling up and down process is essentially the migration of Region Peers among nodes. To migrate a Region Peer from one Write Node to another for management, the target Write node only needs to download a small amount of metadata about this Region from the latest files uploaded to S3 by the Region Peer's original Write Node, and then synchronizes the latest Region updates from TiKV to catch up with the Region Leader's progress. This completes the migration of Region Peers.
```

- TiFlash Compute Node

    The Compute Node executes query requests sent from a TiDB node. It first accesses a Write Node to obtain data snapshots, and then reads the latest data (that is, the data not been uploaded to S3 yet) from the Write Node and most of the remaining data from S3.

    The Compute Node uses local disks (usually NVMe SSDs) as a cache for data files to avoid repeatedly reading the same data from remote locations (Write Nodes or S3) and improve query performance.

    The Compute Node is stateless and its scaling speed is at a second level. You can use this feature to reduce costs as follows:

    - When the query workload is low, reduce the number of Compute Nodes to save costs. When there are no queries, you can even stop all Compute Nodes.
    - When the query workload increases, quickly increase the number of Compute Nodes to ensure query performance.

> ## Use cases

Reviewer 2: ```suggestion
## Scenarios
```

    ```

2. Prepare a TiFlash topology configuration file, such as `scale-out.topo.yaml`, with the following configuration:

    ```yaml
    tiflash_servers:
      # In the TiFlash topology configuration file, the `storage.s3` configuration indicates that the disaggregated storage and compute architecture is used for deployment.
      # If `flash.disaggregated_mode: tiflash_compute` is configured for a node, it is a Compute Node; otherwise, it is a Write Node.

      # 172.31.8.1~2 are TiFlash Write Nodes
      - host: 172.31.8.1
        config:
          flash.disaggregated_mode: tiflash_write               # This is a Write Node
          storage.s3.endpoint: http://s3.{region}.amazonaws.com # S3 endpoint address
          storage.s3.bucket: mybucket                           # TiFlash stores all data in this bucket
          storage.s3.root: /cluster1_data                       # Root directory where data is stored in the S3 bucket
          storage.s3.access_key_id: {ACCESS_KEY_ID}             # Access S3 with ACCESS_KEY_ID
          storage.s3.secret_access_key: {SECRET_ACCESS_KEY}     # Access S3 with SECRET_ACCESS_KEY

>           storage.main.dir: ["/data1/tiflash/data"]             # Local data directory of the Write Node, the same configuration way as that of the coupled storage and compute architecture

Author: ```suggestion
          storage.main.dir: ["/data1/tiflash/data"]             # Local data directory of the Write Node. Configure it in the same way as the directory configuration of the coupled storage and compute architecture
```

      - host: 172.31.8.2
        config:
          flash.disaggregated_mode: tiflash_write               # This is a Write Node
          storage.s3.endpoint: http://s3.{region}.amazonaws.com # S3 endpoint address
          storage.s3.bucket: mybucket                           # TiFlash stores all data in this bucket
          storage.s3.root: /cluster1_data                       # Root directory where data is stored in the S3 bucket
          storage.s3.access_key_id: {ACCESS_KEY_ID}             # Access S3 with ACCESS_KEY_ID
          storage.s3.secret_access_key: {SECRET_ACCESS_KEY}     # Access S3 with SECRET_ACCESS_KEY

>           storage.main.dir: ["/data1/tiflash/data"]             # Local data directory of the Write Node, the same configuration way as that of the coupled storage and compute architecture

Author: ```suggestion
          storage.main.dir: ["/data1/tiflash/data"]             # Local data directory of the Write Node. Configure it in the same way as the directory configuration of the coupled storage and compute architecture
```

      # 172.31.9.1~2 are TiFlash Compute Nodes
      - host: 172.31.9.1
        config:
          flash.disaggregated_mode: tiflash_compute             # This is a Compute Node
          storage.s3.endpoint: http://s3.{region}.amazonaws.com # S3 endpoint address
          storage.s3.bucket: mybucket                           # TiFlash stores all data in this bucket
          storage.s3.root: /cluster1_data                       # Root directory where data is stored in the S3 bucket
          storage.s3.access_key_id: {ACCESS_KEY_ID}             # Access S3 with ACCESS_KEY_ID
          storage.s3.secret_access_key: {SECRET_ACCESS_KEY}     # Access S3 with SECRET_ACCESS_KEY

>           storage.main.dir: ["/data1/tiflash/data"]             # Local data directory of the Compute Node, the same configuration way as that of the coupled storage and compute architecture

Author: ```suggestion
          storage.main.dir: ["/data1/tiflash/data"]             # Local data directory of the Compute Node. Configure it in the same way as the directory configuration of the coupled storage and compute architecture
```

          storage.remote.cache.dir: /data1/tiflash/cache        # Local data cache directory of the Compute Node
          storage.remote.cache.capacity: 858993459200           # 800 GiB
      - host: 172.31.9.2
        config:
          flash.disaggregated_mode: tiflash_compute             # This is a Compute Node
          storage.s3.endpoint: http://s3.{region}.amazonaws.com # S3 endpoint address
          storage.s3.bucket: mybucket                           # TiFlash stores all data in this bucket
          storage.s3.root: /cluster1_data                       # Root directory where data is stored in the S3 bucket
          storage.s3.access_key_id: {ACCESS_KEY_ID}             # Access S3 with ACCESS_KEY_ID
          storage.s3.secret_access_key: {SECRET_ACCESS_KEY}     # Access S3 with SECRET_ACCESS_KEY

>           storage.main.dir: ["/data1/tiflash/data"]             # Local data directory of the Compute Node, the same configuration way as that of the coupled storage and compute architecture

Reviewer 2: ```suggestion
          storage.main.dir: ["/data1/tiflash/data"]             # Local data directory of the Compute Node. Configure it in the same way as the configuration of the coupled storage and compute architecture
```

          storage.remote.cache.dir: /data1/tiflash/cache        # Local data cache directory of the Compute Node
          storage.remote.cache.capacity: 858993459200           # 800 GiB
    ```

>     * Note that the above `ACCESS_KEY_ID` and `SECRET_ACCESS_KEY` are directly written in the configuration file. You can also choose to configure them separately using environment variables. If both ways are configured. the priority of environment variables is higher than that of configuration files.

Reviewer 2: ```suggestion
    * Note that the above `ACCESS_KEY_ID` and `SECRET_ACCESS_KEY` are directly written in the configuration file. You can also choose to configure them separately using environment variables. If both ways are configured, the environment variables have higher priority.
```

    1. Open the TiDB configuration file in edit mode:

          ```shell
          tiup cluster edit-config mycluster
          ```

    2. Add the following configuration items to the TiDB configuration file:

        ```shell
        server_configs:
        tidb:
        disaggregated-tiflash: true   # Query TiFlash using the disaggregated storage and compute architecture
        ```

    3. Restart TiDB:

        ```shell
        tiup cluster reload mycluster -R tidb
        ```

## Restrictions

- TiFlash does not support in-place switching between the **disaggregated storage and compute architecture** and the **coupled storage and compute architecture**. Before switching architectures, you must remove all existing TiFlash nodes.
- After the migration from one architecture to another, all TiFlash data needs to be replicated again.
- Only TiFlash nodes with the same architecture are allowed in the same TiDB cluster. Two architectures cannot coexist in one cluster.
- The disaggregated storage and compute architecture only supports object storage using the S3 API, while the coupled storage and compute architecture only supports local storage.

> - When using S3 storage, TiFlash nodes cannot obtain the keys of files not on their own nodes, so the [Encryption at Rest](/encryption-at-rest.md) feature is cannot be used.

Reviewer 2: ```suggestion
- When using S3 storage, TiFlash nodes cannot obtain the keys of files not on their own nodes, so the [Encryption at Rest](/encryption-at-rest.md) feature cannot be used.
```

## Architecture overview

![TiFlash Write and Compute Separation Architecture](/media/tiflash/tiflash-s3.png)

In the disaggregated storage and compute architecture, different functionalities of the TiFlash process are divided and allocated to two types of nodes: the Write Node and the Compute Node. These two types of nodes can be deployed separately and scaled independently, which means that you can decide the number of Write Nodes and Compute Nodes to be deployed as needed.

- TiFlash Write Node

    The Write Node receives Raft logs data from TiKV, converts the data into the columnar format, and periodically packages and uploads all the updated data within a certain period to S3. In addition, the Write Node manages the data on S3, such as continuously organizing data to improve query performance and deleting useless data.

    The Write Node uses local disks (usually NVMe SSDs) to cache the latest written data to avoid excessive use of memory.

>     The Write Node has a faster scaling speed than the TiFlash node in the coupled storage and compute architecture. That is, after adding or removing Write Nodes, data can reach balance faster among Write Nodes. The reason is that the Write Node stores all the data in S3 and only needs to store a small amount of data locally at runtime. The scaling up and down process is essentially the migration of Region Peers among nodes. To migrate a Region Peer from one Write Node to another for management, the target Write node only needs to download a small amount of metadata about this Region from the latest files uploaded to S3 by the Region Peer's original Write Node, and then synchronizes the latest Region updates from TiKV to catch up with the Region Leader's progress. This completes the migration of Region Peers.

Author: ```suggestion
    The Write Node has a faster scaling speed than the TiFlash node in the coupled storage and compute architecture. That is, after adding or removing Write Nodes, data can reach balance faster among Write Nodes. The reason is that the Write Node stores all the data in S3 and only needs to store a small amount of data locally at runtime. The scaling up and down process is essentially the migration of Region Peers among nodes. When migrating a Region Peer from one Write Node to another for management, the target Write node only needs to download a small amount of metadata about this Region from the latest files uploaded to S3 by the Region Peer's original Write Node, and then synchronize the latest Region updates from TiKV. This catches up with the Region Leader's progress and completes the migration of Region Peers.
```

>     The Write Node has a faster scaling speed than the TiFlash node in the coupled storage and compute architecture. That is, after adding or removing Write Nodes, data can reach balance faster among Write Nodes. The reason is that the Write Node stores all the data in S3 and only needs to store a small amount of data locally at runtime. The scaling up and down process is essentially the migration of Region Peers among nodes. When migrating a Region Peer from one Write Node to another for management, the target Write node only needs to download a small amount of metadata about the Region from the latest files uploaded to S3 by the Region Peer's original Write Node, and then synchronize the latest Region updates from TiKV. Then, the target Write node can catch up with the Region Leader's progress and complete the migration of the Region Peers.

Author: ```suggestion
    The Write Node has a faster scaling speed than the TiFlash node in the coupled storage and compute architecture. That is, after adding or removing Write Nodes, data can reach balance faster among Write Nodes. The reason is that the Write Node stores all the data in S3 and only needs to store a small amount of data locally at runtime. The scaling up and down process is essentially the migration of Region Peers among nodes. When migrating a Region Peer from one Write Node to another for management, the target Write node only needs to download a small amount of metadata about the Region from the latest files uploaded to S3 by the Region Peer's original Write Node, and synchronizes the latest Region updates from TiKV. Then, the target Write node can catch up with the Region Leader's progress and complete the migration of the Region Peers.
```
